Source/plrmsg.cpp

```diff
@@ -110,26 +128,70 @@ void InitPlrMsg()
 void DrawPlrMsg(const Surface &out)
 {
 	int x = 10;
-	int y = 58;
+	int y = PANEL_TOP - 32;
 	int width = gnScreenWidth - 20;
+	int vislines;
+	int chatlines;
 	_plrmsg *pMsg;
 
-	if (chrflag || QuestLogIsOpen) {
+	int oneLine = 15;
+	int twoLine = 30;
+	int threeLine = 45;
+
+	if (!talkflag && (chrflag || QuestLogIsOpen/* || stashflag*/)) {
 		x += GetLeftPanel().position.x + GetLeftPanel().size.width;
 		width -= GetLeftPanel().size.width;
 	}
-	if (invflag || sbookflag)
+	if (!talkflag && (invflag || sbookflag))
 		width -= gnScreenWidth - GetRightPanel().position.x;
 
 	if (width < 300)
 		return;
 
+	if (talkflag) {
+		vislines = PMSG_COUNT;
+	} else {
+		vislines = 3;
+	}
+
 	pMsg = plr_msgs;
-	for (int i = 0; i < PMSG_COUNT; i++) {
-		if (pMsg->str[0] != '\0')
-			PrintChatMessage(out, x, y, width, pMsg->str, TextColorFromPlayerId[pMsg->player]);
-		pMsg++;
-		y += 35;
+
+	for (int i = 0; i < vislines; i++) {
+		if (pMsg->str[0] != '\0') {
+			if (talkflag) {
+				if (GetChatLines(width, pMsg->str) == 1) {
+					DrawHalfTransparentRectTo(out, x - 3, y, width + 6, oneLine);
+				} else if (GetChatLines(width, pMsg->str) == 2) {
+					DrawHalfTransparentRectTo(out, x - 3, y, width + 6, twoLine);
+				} else {
+					DrawHalfTransparentRectTo(out, x - 3, y, width + 6, threeLine);
+				}
+				PrintChatMessage(out, x, y, width, pMsg->str, TextColorFromPlayerId[pMsg->player]);
+				if (pMsg->player != MAX_PLRS)
+					PrintChatMessage(out, x, y, width, pMsg->name, TextColorFromPlayerId[MAX_PLRS]);
+			} else if (SDL_GetTicks() - pMsg->time < 10000) {
+				if (GetChatLines(width, pMsg->str) == 1) {
+					DrawHalfTransparentRectTo(out, x - 3, y, width + 6, oneLine);
+				} else if (GetChatLines(width, pMsg->str) == 2) {

```

Comment:
`GetChatLines` looks pretty expensive and there really isn't a good reason to call int multiple times:

```suggestion
			chatlines = GetChatLines(width, pMsg->str);
			if (talkflag) {
				if (chatlines == 1) {
					DrawHalfTransparentRectTo(out, x - 3, y, width + 6, oneLine);
				} else if (chatlines == 2) {
					DrawHalfTransparentRectTo(out, x - 3, y, width + 6, twoLine);
				} else {
					DrawHalfTransparentRectTo(out, x - 3, y, width + 6, threeLine);
				}
				PrintChatMessage(out, x, y, width, pMsg->str, TextColorFromPlayerId[pMsg->player]);
				if (pMsg->player != MAX_PLRS)
					PrintChatMessage(out, x, y, width, pMsg->name, TextColorFromPlayerId[MAX_PLRS]);
			} else if (SDL_GetTicks() - pMsg->time < 10000) {
				if (chatlines == 1) {
					DrawHalfTransparentRectTo(out, x - 3, y, width + 6, oneLine);
				} else if (chatlines == 2) {
```